proposal.md

```diff
@@ -0,0 +1,186 @@
+/usr/src/app/startup/init.js - in “app” package scope
+/usr/src/app/node_modules/   - in “app” package scope
+
+/usr/src/app/node_modules/sinon/package.json - in “sinon” package scope
+/usr/src/app/node_modules/sinon/lib/sinon.js - in “sinon” package scope
+```
+
+The `package.json` files here each create a new package scope for the folder they’re in, and all subfolders down until another `package.json` file creates a new scope below. Thus `/usr/src/app/` and `/usr/src/app/node_modules/sinon/` are each package boundaries, with different package scopes above and below.
+
+Within a package scope, `package.json` files can hold metadata and configuration for the files contained within that scope. This proposal uses this concept to allow `package.json` values to opt in a package into being interpreted as ESM.
+
+Package scopes apply only to files. While a user may execute code via `node --eval` from within a working directory of a particular package scope, that package scope has no effect on the source code passed in via `node --eval`.
+
+## Package scope algorithm
+
+After Node uses the algorithm in the [resolver specification][nodejs/ecmascript-modules:esm.md#resolver-algorithm] to locate a source code file to load, Node must then decide whether to load it as ESM or as CommonJS. If the source code file has an `.mjs` extension, Node is done; the file is loaded as ESM. Otherwise, Node needs to find the `package.json` governing the package scope that the file is in. That package scope algorithm goes as follows:
+
+```
+If the file is a package entry point
+    And the package’s package.json says to load JavaScript as ESM
+        Load the file as ESM.
+    Else
+        Load the file as CommonJS.
+Else
+    If there is a package.json in the folder where the file is
+        And the package.json says to load JavaScript as ESM
+            Load the file as ESM.
+        Else
+            Load the file as CommonJS.
+    Else
+        Go into the parent folder and look for a package.json there
+        Repeat until we either find a package.json or hit the file system root
+        If we found a package.json
+            And the package.json says to load JavaScript as ESM
+                Load the file as ESM.
+            Else
+                Load the file as CommonJS.
+        Else we reach the file system root without finding a package.json
+            Load the file as CommonJS if initial entry point, ESM otherwise.
+```
+
+The last case, when we reach the file system root without finding a `package.json`, covers files that are outside any package scope. If such files are the initial entry point, e.g. `node /usr/local/bin/backup.js`, they are parsed as CommonJS for backward compatibility. If such files are referenced via an `import` statement or `import()` expression, they are parsed as ESM.
+
+## `package.json` configuration
+
+> Coming soon!
+
+## `import` statements and `import()` expressions
+
+### `import` specifiers
```

Comment:
```suggestion
## Module Specifiers
```

Do we want to separate this out into its own H2 section? I feel that it is on the same level as 1) Package Configuration and 2) Import Statements and Expressions

Reply:
So an `import` statement is a full line like `import _ from 'lodash'`. Within that statement, the `'lodash'` part by itself is a specifier. Within `import('./file.js')`, the `'./file.js'` part is a specifier. This list is a list of the various types of specifiers that `import` statements and `import()` expressions can include. That’s why I consider it a part of the section on `import` statements and expressions.